consensus/consensus.go

```diff
@@ -1,103 +1,237 @@
 package consensus
 
 import (
-	"math/big"
+	"crypto/ecdsa"
+	"math"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/ubclaunchpad/cumulus/blockchain"
+	c "github.com/ubclaunchpad/cumulus/common/constants"
 )
 
-// MinedBlockCode is returned from ValidateMinedBlock
-type MinedBlockCode uint32
-
-const (
-	// ValidNewBlock is returned when a mined block is valid
-	ValidNewBlock MinedBlockCode = iota
-	// BadBlock is returned when an invalid block was mined
-	BadBlock MinedBlockCode = iota
-	// BadNonce is returned when the hash of the block is not less than the
-	// target
-	BadNonce MinedBlockCode = iota
-	// BadTarget is returned when the target that was used while mining is
-	// not equal to the current network target
-	BadTarget MinedBlockCode = iota
-	// BadBlockReward is returned when the block's rewards is not equal to the
-	// current reward of the network
-	BadBlockReward MinedBlockCode = iota
-	// BadCloudBase is returned when the address of the recipient of the
-	// CloudBase transaction is not valid
-	BadCloudBase MinedBlockCode = iota
-)
+// VerifyTransaction tests whether a transaction valid.
+func VerifyTransaction(bc *blockchain.BlockChain,
+	t *blockchain.Transaction) (bool, TransactionCode) {
 
-const (
-	// blockRewardHalvingRate is the number of blocks that need to be mined
-	// before the blockReward is halved
-	blockRewardHalvingRate int = 210000
-)
+	// Check if the transaction is equal to nil
+	if t == nil {
+		return false, NilTransaction
+	}
 
-var (
-	// BlockReward is the current reward for mining a block
-	BlockReward uint64 = 25
-	// CurrentDifficulty is the current hashing difficulty of the network
-	CurrentDifficulty = blockchain.MinTarget
-)
+	// Find the transaction input in the chain (by hash)
+	var input *blockchain.Transaction
+	input = bc.GetInputTransaction(t)
+	if input == nil || blockchain.HashSum(input) != t.Input.Hash {
+		return false, NoInputTransaction
+	}
+
+	// Check that output to sender in input is equal to outputs in t
+	if !input.InputsEqualOutputs(t) {
+		return false, Overspend
+	}
 
-// HalveReward halves the current blockReward if the size of the BlockChain is a
-// multiple of the blockRewardHalvingRate
-func HalveReward(bc *blockchain.BlockChain) {
-	if len(bc.Blocks)%blockRewardHalvingRate == 0 {
-		BlockReward /= 2
+	// Verify signature of t
+	hash := blockchain.HashSum(t.TxBody)
+	if !ecdsa.Verify(t.Sender.Key(), hash.Marshal(), t.Sig.R, t.Sig.S) {
+		return false, BadSig
 	}
+
+	// Test if identical transaction already exists in chain.
+	end := uint32(len(bc.Blocks))
+	start := t.Input.BlockNumber
+	if exists, _, _ := bc.ContainsTransaction(t, start, end); exists {
```

Comment:
All this `ContainsTransaction` bit does is check if any transaction in the chain hashes to `HashSum(t)`. Literally `HashSum(t) == HashSum(t0)` for all intervening transactions. 

If Alice has 10 and sends Bob 10 then sends Bob 9.99, I don't think we catch it because the outputs look different and that changes the hashes. We may catch it elsewhere, but I can't see it 🤔